src/client/address_book.cc

```diff
@@ -55,221 +55,292 @@
+    if (m_SubscriptionIsLoaded
+        && !m_SubscriberIsDownloading
+        && m_SharedLocalDestination->IsReady()) {
+      // Pick a random subscription if subscriptions count is > 0
+      // At first this may look questionable but, by this point,
+      // we're guaranteed to have at least one subscriber
+      auto publisher =
+        i2p::crypto::RandInRange<std::size_t>(0, m_Subscribers.size() - 1);
+      m_SubscriberIsDownloading = true;
+      m_Subscribers.at(publisher)->DownloadSubscription();
+    } else {
+      if (!m_SubscriptionIsLoaded) {
+        // If subscription not available, will attempt download with subscriber
+        LoadSubscriptionFromPublisher();
+      }
+      // Try again after timeout
+      m_SubscriberUpdateTimer->expires_from_now(
+          boost::posix_time::minutes(
+              static_cast<std::uint16_t>(SubscriberTimeout::InitialRetry)));
+      m_SubscriberUpdateTimer->async_wait(
+          std::bind(
+              &AddressBook::SubscriberUpdateTimer,
+              this,
+              std::placeholders::_1));
+    }
   } else {
     LogPrint(eLogError,
-        "AddressBookFilesystemStorage: can't open file ", filename);
+        "AddressBook: SubscriberUpdateTimer() exception: ", ecode.message());
   }
 }
 
-void AddressBookFilesystemStorage::RemoveAddress(
-    const i2p::data::IdentHash& ident) {
-  auto filename = GetPath() / (ident.ToBase32() + ".b32");
-  if (boost::filesystem::exists(filename))
-    boost::filesystem::remove(filename);
-}
-
-int AddressBookFilesystemStorage::Load(
-    std::map<std::string, i2p::data::IdentHash>& addresses) {
-  int num = 0;
-  auto filename = GetPath() / "addresses.csv";
-  std::ifstream f(filename.string(), std::ofstream::in);  // in text mode
-  if (f.is_open()) {
-    addresses.clear();
-    while (!f.eof()) {
-      std::string s;
-      getline(f, s);
-      if (!s.length())
-        continue;  // skip empty line
-      size_t pos = s.find(',');
-      if (pos != std::string::npos) {
-        std::string name = s.substr(0, pos++);
-        std::string addr = s.substr(pos);
-        i2p::data::IdentHash ident;
-        ident.FromBase32(addr);
-        addresses[name] = ident;
-        num++;
+void AddressBook::LoadSubscriptionFromPublisher() {
+  // First, ensure we have a storage instance ready
+  if (!m_Storage)
+    m_Storage = GetNewStorageInstance();
+  // If so, see if we have addresses from subscription already saved
+  if (m_Storage->Load(m_Addresses)) {
+    // If so, we don't need to download from a publisher
+    m_SubscriptionIsLoaded = true;
+    return;
+  }
+  // If available, load default subscription from file
+  auto filename = GetDefaultSubscriptionFilename();
+  std::ifstream file(
+      i2p::util::filesystem::GetFullPath(filename),
+      std::ofstream::in);
+  LogPrint(eLogInfo, "AddressBook: loading subscription ", filename);
+  if (file.is_open()) {  // Open subscription, validate, and save to storage
```

Comment:
@anonimal Maybe we should use `operator bool` to check whether the file is OK for reading. `is_open()` only checks if there is an associated file.